File: longest_common_substring.py

def longestCommSubStr(str1, str2):
    # Create the string we want to return.
    final_lcs = ''

    # Length of the first string.
    len1 = len(str1)

    # Length of the second string.
    len2 = len(str2)

    # Iterate through the positions of each character in the columns (first string characters) from left to right
    for i in range(len1):

        # We compare that position's character with each row position's character (second string characters)
        for j in range(len2):
            # Create a variable to mark the positions of each character across the first string columns and down the second string rows.
            pos = 0

            # Create a variable as a temporary longest common substring that we can add characters to if conditions are met. 
            temp_lcs = ''

            # While the first string position is less than the first string length
            # and the second string position is less than the second string length
            # and the letter at the first string position is equal to character at the second string position,
            # this means the characters match and we need to add that character to our string we want returned.
            while ((i + pos < len1) and (j + pos < len2) and str1[i + pos] == str2[j + pos]):

                # Our temporary longest common substring is equal to its current value plus the matched letter.
                temp_lcs += str2[j + pos]

                # 
                if str1[i + pos] == str2[j + pos]:
                    exit

                # We set the position to the next position 
                pos += 1
                
            # If the length of our temporary string is greater the length of the final string we want returned...
            if (len(temp_lcs) > len(final_lcs)):

                #...we set our final string equal to the temorary one.  
                # Keep in mind, we declared our final_lcs outside the loop, so only the 
                # resetting the temp_lcs for ever letter of the first string 
                # is not resetting the final_lcs to an empty string.
                final_lcs = temp_lcs
                
    return final_lcs

lcs = longestCommSubStr("nailed it", "snailed it")
print lcs
print len(lcs)

File: LongestCommonSubstring.java
package com.stephenjohnny.tools;

public class LongestCommonSubstring {

    public static void main(String[] args) {
        longestSubstring("abcdefg", "andabcdemore");
    }
    public static int longestSubstring(String s1, String s2) {

        // We create a 'table' that looks like this:
        //   a b c d e f g h i j 
        // o 0 0 0 0 0 0 0 0 0 0 
        // r 0 0 0 0 0 0 0 0 0 0 
        // a 1 0 0 0 0 0 0 0 0 0 = new maxLength = 1
        // b 0 2 0 0 0 0 0 0 0 0 = new maxLength = 2 (because adding 1 to the value of [0][2] equals 2)
        // c 0 0 3 0 0 1 0 0 0 0 = new maxLength = 3 (because adding 1 to the value of [1][3] equals 3)
        // i 0 0 0 0 0 0 0 0 1 0
        // n 0 0 0 0 0 0 0 0 0 0 
        // s 0 0 0 0 0 0 0 0 0 0 
        // t 0 0 0 0 0 0 0 0 0 0 
        // e 0 0 0 0 1 0 0 0 0 0 
        // a 1 0 0 0 0 0 0 0 0 0
        // d 0 0 0 1 0 0 0 0 0 0 

        // This is set to '0' by default
        int[][] num = new int[s1.length()][s2.length()];
        
        // This will be the placeholder for the value we need to return
        int maxLength = 0;

        for (int i = 0; i < s1.length(); i++) {
            
            for (int j = 0; j < s2.length(); j++) {
                
                // If the index character of string 1 equals the index character of string 2
                if (s1.charAt(i) == s2.charAt(j)) {

                    if ((i == 0) || (j == 0))
                        num[i][j] = 1;
                    else                 
                        // We add 1 to the value of the previous index positions
                        num[i][j] = 1 + num[i - 1][j - 1];
                    
                    // If we've now added a value and it's greater than the max placeholder
                    // We assign it to the new max placeholder
                    if (num[i][j] > maxLength) {
                        maxLength = num[i][j];
                    }
                }
            }
        }
        System.out.println("The longest common substring contains: " + maxLength + " characters.");
        return maxLength;
    }
}
